File: common/src/main/java/generations/gg/generations/core/generationscore/common/SimpleItemStorage.java
package generations.gg.generations.core.generationscore.common;

import earth.terrarium.common_storage_lib.context.ItemContext;
import earth.terrarium.common_storage_lib.data.DataManager;
import earth.terrarium.common_storage_lib.item.util.ItemStorageData;
import earth.terrarium.common_storage_lib.resources.ResourceStack;
import earth.terrarium.common_storage_lib.resources.item.ItemResource;
import earth.terrarium.common_storage_lib.storage.base.CommonStorage;
import earth.terrarium.common_storage_lib.storage.base.UpdateManager;
import earth.terrarium.common_storage_lib.storage.util.TransferUtil;
import net.minecraft.core.NonNullList;
import net.minecraft.core.component.DataComponentType;
import org.jetbrains.annotations.NotNull;

import java.util.Objects;
import java.util.function.Predicate;

public class SimpleItemStorage implements CommonStorage<ItemResource>, UpdateManager<ItemStorageData> {
    protected final NonNullList<SimpleItemSlot> slots;
    private final Runnable onUpdate;
    private final Runnable save;

    public SimpleItemStorage(int size) {
        this.slots = NonNullList.withSize(size, new SimpleItemSlot(this::update));
        this.onUpdate = () -> {
        };
        this.save = () -> {
        };
    }

    public SimpleItemStorage(ItemContext context, DataComponentType<ItemStorageData> componentType, int size) {
        this.slots = NonNullList.withSize(size, new SimpleItemSlot(this::update));
        Objects.requireNonNull(context);
        this.onUpdate = context::updateAll;
        this.save = () -> {
            context.set(componentType, ItemStorageData.of(this));
        };
        if (context.getResource().has(componentType)) {
            this.readSnapshot(context.getResource().get(componentType));
        }

    }

    public SimpleItemStorage(Object entityOrBlockEntity, DataManager<ItemStorageData> dataManager, int size) {
        this.slots = NonNullList.withSize(size, new SimpleItemSlot(this::update));
        this.onUpdate = () -> {
            dataManager.set(entityOrBlockEntity, ItemStorageData.of(this));
        };
        this.save = () -> {
        };
        this.readSnapshot(dataManager.get(entityOrBlockEntity));
    }

    public SimpleItemStorage filter(int slot, Predicate<ItemResource> predicate) {
        this.slots.set(slot, new SimpleItemSlot.Filtered(this::update, predicate));
        return this;
    }

    public int size() {
        return this.slots.size();
    }

    public @NotNull SimpleItemSlot get(int index) {
        return this.slots.get(index);
    }

    public ItemStorageData createSnapshot() {
        return ItemStorageData.of(this);
    }

    public void readSnapshot(ItemStorageData snapshot) {
        for(int i = 0; i < this.slots.size() && i < snapshot.stacks().size(); ++i) {
            ((SimpleItemSlot)this.slots.get(i)).readSnapshot((ResourceStack)snapshot.stacks().get(i));
        }

    }

    public void update() {
        this.onUpdate.run();
    }

    public long insert(ItemResource resource, long amount, boolean simulate) {
        return TransferUtil.insertSlots(this, resource, amount, simulate);
    }

    public long extract(ItemResource resource, long amount, boolean simulate) {
        return TransferUtil.extractSlots(this, resource, amount, simulate);
    }
}

File: common/src/main/java/generations/gg/generations/core/generationscore/common/world/data/SimpleItemStorage.kt
package generations.gg.generations.core.generationscore.common.world.data

import earth.terrarium.common_storage_lib.context.ItemContext
import earth.terrarium.common_storage_lib.data.DataManager
import earth.terrarium.common_storage_lib.item.impl.SimpleItemSlot
import earth.terrarium.common_storage_lib.item.util.ItemStorageData
import earth.terrarium.common_storage_lib.resources.ResourceStack
import earth.terrarium.common_storage_lib.resources.item.ItemResource
import earth.terrarium.common_storage_lib.storage.base.CommonStorage
import earth.terrarium.common_storage_lib.storage.base.UpdateManager
import earth.terrarium.common_storage_lib.storage.util.TransferUtil
import net.minecraft.core.NonNullList
import net.minecraft.core.component.DataComponentType
import net.minecraft.world.entity.Entity
import net.minecraft.world.level.block.entity.BlockEntity
import java.util.*
import java.util.function.Predicate

class SimpleItemStorage : CommonStorage<ItemResource>, UpdateManager<ItemStorageData> {
    protected val slots: NonNullList<SimpleItemSlot>
    private val onUpdate: Runnable
    private val save: Runnable

    constructor(size: Int) {
        this.slots = NonNullList.withSize(size, SimpleItemSlot { this.update() })
        this.onUpdate = Runnable {}
        this.save = Runnable {}
    }

    constructor(context: ItemContext, componentType: DataComponentType<ItemStorageData>, size: Int) {
        this.slots = NonNullList.withSize(size, SimpleItemSlot { this.update() })
        Objects.requireNonNull(context)
        this.onUpdate = Runnable { context.updateAll() }
        this.save = Runnable {
            context.set(
                componentType, ItemStorageData.of(
                    this
                )
            )
        }
        if (context.resource.has(componentType)) {
            this.readSnapshot(context.resource.get(componentType)!!)
        }
    }

    constructor(entityOrBlockEntity: BlockEntity, dataManager: DataManager<ItemStorageData>, size: Int) {
        this.slots = NonNullList.withSize(size, SimpleItemSlot { this.update() })
        this.onUpdate = Runnable {
            dataManager[entityOrBlockEntity] = ItemStorageData.of(this)
        }
        this.save = Runnable {}
        this.readSnapshot((dataManager[entityOrBlockEntity] as ItemStorageData))
    }

    constructor(entityOrBlockEntity: Entity, dataManager: DataManager<ItemStorageData>, size: Int) {
        this.slots = NonNullList.withSize(size, SimpleItemSlot { this.update() })
        this.onUpdate = Runnable {
            dataManager[entityOrBlockEntity] = ItemStorageData.of(this)
        }
        this.save = Runnable {}
        this.readSnapshot((dataManager[entityOrBlockEntity] as ItemStorageData))
    }

    fun filter(slot: Int, predicate: Predicate<ItemResource?>?): SimpleItemStorage {
        slots[slot] =
            SimpleItemSlot.Filtered(
                { this.update() },
                predicate
            )
        return this
    }

    override fun size(): Int {
        return slots.size
    }

    override fun get(index: Int): SimpleItemSlot {
        return slots[index]
    }

    override fun createSnapshot(): ItemStorageData {
        return ItemStorageData.of(this)
    }

    override fun readSnapshot(snapshot: ItemStorageData) {
        var i = 0
        while (i < slots.size && i < snapshot.stacks().size) {
            slots[i].readSnapshot(snapshot.stacks()[i])
            ++i
        }
    }

    override fun update() {
        onUpdate.run()
    }

    override fun insert(resource: ItemResource?, amount: Long, simulate: Boolean): Long {
        return TransferUtil.insertSlots(this, resource, amount, simulate)
    }

    override fun extract(resource: ItemResource?, amount: Long, simulate: Boolean): Long {
        return TransferUtil.extractSlots(this, resource, amount, simulate)
    }
}
